### PR TITLE
MINOR: Broker should disallow downconversion of transactional/idempotent records

### PR DESCRIPTION
**Pull request by hachikuji**

### CODE REVIEW

File: core/src/main/scala/kafka/log/LogValidator.scala [hunk 1]
```diff
@@ -78,6 +78,12 @@ private[kafka] object LogValidator extends Logging {
       if (batch.isControlBatch)
         throw new InvalidRecordException("Clients are not allowed to write control records")
     }
+
+    if (batch.isTransactional && toMagic < RecordBatch.MAGIC_VALUE_V2)
+      throw new UnsupportedForMessageFormatException(s"Transactional records cannot be used with magic version $toMagic")
```
Comment (guozhangwang) [inline core/src/main/scala/kafka/log/LogValidator.scala:83]: Is this ever expected to happen? If it is then we should not throw an exception to fail the broker but return an error code back to client.

Comment (guozhangwang) [reply core/src/main/scala/kafka/log/LogValidator.scala:83]: What I'm wondering if users made a mistake in operations while setting `messageFormatVersion`, in that case should we throw an exception to stop the broker upon receiving a first transactional message?

Comment (hachikuji) [reply core/src/main/scala/kafka/log/LogValidator.scala:83]: The `UnsupportedForMessageFormatException` should get caught by the request handler and returned as an error code to the user, right? That is why I updated the error codes for the `ProduceRequest`.

Comment (guozhangwang) [reply core/src/main/scala/kafka/log/LogValidator.scala:83]: That's right.

**Comment by guozhangwang** [comment]
LGTM.

**Pull request closed without merging**